File: app/src/main/java/com/kispoko/tome/engine/programming/variable/NumberVariable.java

package com.kispoko.tome.engine.programming.variable;


import com.kispoko.tome.ApplicationFailure;
import com.kispoko.tome.error.InvalidCaseError;
import com.kispoko.tome.error.UnknownVariantError;
import com.kispoko.tome.exception.InvalidDataException;
import com.kispoko.tome.engine.programming.program.invocation.Invocation;
import com.kispoko.tome.engine.programming.summation.Summation;
import com.kispoko.tome.engine.programming.summation.SummationException;
import com.kispoko.tome.engine.refinement.RefinementId;
import com.kispoko.tome.exception.UnionException;
import com.kispoko.tome.util.EnumUtils;
import com.kispoko.tome.util.database.DatabaseException;
import com.kispoko.tome.util.database.sql.SQLValue;
import com.kispoko.tome.util.model.Model;
import com.kispoko.tome.util.value.ModelFunctor;
import com.kispoko.tome.util.value.PrimitiveFunctor;
import com.kispoko.tome.util.yaml.Yaml;
import com.kispoko.tome.util.yaml.YamlException;
import com.kispoko.tome.util.yaml.error.InvalidEnumError;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;



/**
 * Number Variable
 */
public class NumberVariable extends Variable implements Model, Serializable
{

    // PROPERTIES
    // ------------------------------------------------------------------------------------------

    // > Model
    // ------------------------------------------------------------------------------------------

    private UUID                        id;


    // > Functors
    // ------------------------------------------------------------------------------------------

    private PrimitiveFunctor<String>    name;

    private PrimitiveFunctor<Integer>   integerValue;
    private ModelFunctor<Invocation>    invocationValue;
    private ModelFunctor<Summation>     summation;

    private PrimitiveFunctor<Kind>      kind;
    private ModelFunctor<RefinementId>  refinementId;
    private PrimitiveFunctor<String[]>  tags;


    // > Internal
    // ------------------------------------------------------------------------------------------

    private ReactiveValue<Integer>      reactiveValue;


    // CONSTRUCTORS
    // ------------------------------------------------------------------------------------------

    public NumberVariable()
    {
        super();

        this.id              = null;

        this.name            = new PrimitiveFunctor<>(null, String.class);

        this.integerValue    = new PrimitiveFunctor<>(null, Integer.class);
        this.invocationValue = ModelFunctor.empty(Invocation.class);
        this.summation       = ModelFunctor.empty(Summation.class);

        this.kind            = new PrimitiveFunctor<>(null, Kind.class);
        this.refinementId    = ModelFunctor.empty(RefinementId.class);
        this.tags            = new PrimitiveFunctor<>(null, String[].class);

        this.reactiveValue   = null;
    }


    /**
     * Create a Variable. This constructor is private to enforce use of the case specific
     * constructors, so only valid value/type associations can be used.
     * @param id The Model id.
     * @param value The Variable value.
     * @param kind The Number Variable kind.
     */
    private NumberVariable(UUID id,
                           String name,
                           Object value,
                           Kind kind,
                           RefinementId refinementId,
                           List<String> tags)
    {
        super();

        this.id                     = id;

        this.name                   = new PrimitiveFunctor<>(name, String.class);

        this.integerValue           = new PrimitiveFunctor<>(null, Integer.class);
        this.invocationValue = ModelFunctor.full(null, Invocation.class);
        this.summation              = ModelFunctor.full(null, Summation.class);

        this.kind                   = new PrimitiveFunctor<>(kind, Kind.class);
        this.refinementId           = ModelFunctor.full(refinementId, RefinementId.class);

        String[] tagsArray = new String[tags.size()];
        tags.toArray(tagsArray);
        this.tags                   = new PrimitiveFunctor<>(tagsArray, String[].class);

        // Set value according to variable type
        switch (kind)
        {
            case LITERAL:
                this.integerValue.setValue((Integer) value);
                break;
            case PROGRAM:
                this.invocationValue.setValue((Invocation) value);
                break;
            case SUMMATION:
                this.summation.setValue((Summation) value);
                break;
        }

        initialize();
    }


    /**
     * Create a "literal" number variable that contains a value of type Integer.
     * @param id The Model id.
     * @param integerValue The Integer value.
     * @return A new "literal" Integer Variable.
     */
    public static NumberVariable asInteger(UUID id,
                                           String name,
                                           Integer integerValue,
                                           RefinementId refinementId,
                                           List<String> tags)
    {
        return new NumberVariable(id, name, integerValue, Kind.LITERAL, refinementId, tags);
    }


    /**
     * Create a "program" valued variable.
     * @param id The Model id.
     * @param invocation The Invocation value.
     * @return A new "program" variable.
     */
    public static NumberVariable asProgram(UUID id,
                                           String name,
                                           Invocation invocation,
                                           RefinementId refinementId,
                                           List<String> tags)
    {
        return new NumberVariable(id, name, invocation, Kind.PROGRAM, refinementId, tags);
    }


    /**
     * Create the "summation" case.
     * @param id The Model id.
     * @param summation The summation.
     * @return The new number variable as a summation.
     */
    public static NumberVariable asSummation(UUID id,
                                             String name,
                                             Summation summation,
                                             RefinementId refinementId,
                                             List<String> tags)
    {
        return new NumberVariable(id, name, summation, Kind.SUMMATION, refinementId, tags);
    }


    /**
     * Create a new Variable from its Yaml representation.
     * @param yaml The Yaml parser.
     * @return The new Variable.
     * @throws YamlException
     */
    public static NumberVariable fromYaml(Yaml yaml)
                  throws YamlException
    {
        if (yaml.isNull())
            return null;

        UUID         id           = UUID.randomUUID();
        String       name         = yaml.atMaybeKey("name").getString();
        Kind         kind         = Kind.fromYaml(yaml.atKey("type"));
        RefinementId refinementId = RefinementId.fromYaml(yaml.atMaybeKey("refinement"));
        List<String> tags         = yaml.atMaybeKey("tags").getStringList();

        switch (kind)
        {
            case LITERAL:
                Integer integerValue  = yaml.atKey("value").getInteger();
                return NumberVariable.asInteger(id, name, integerValue, refinementId, tags);
            case PROGRAM:
                Invocation invocation = Invocation.fromYaml(yaml.atKey("value"));
                return NumberVariable.asProgram(id, name, invocation, refinementId, tags);
            case SUMMATION:
                Summation summation = Summation.fromYaml(yaml.atKey("value"));
                return NumberVariable.asSummation(id, name, summation, refinementId, tags);
        }

        // CANNOT REACH HERE. If VariableKind is null, an InvalidEnum exception would be thrown.
        return null;
    }


    // API
    // ------------------------------------------------------------------------------------------

    // > Model
    // ------------------------------------------------------------------------------------------

    // ** Id
    // ------------------------------------------------------------------------------------------

    /**
     * Get the model identifier.
     * @return The model UUID.
     */
    public UUID getId()
    {
        return this.id;
    }


    /**
     * Set the model identifier.
     * @param id The new model UUID.
     */
    public void setId(UUID id)
    {
        this.id = id;
    }


    // ** On Load
    // ------------------------------------------------------------------------------------------

    public void onLoad()
    {
        initialize();
    }


    // > State
    // ------------------------------------------------------------------------------------------

    /**
     * The invocation case.
     * @return The invocation.
     */
    public Invocation invocation()
    {
        if (this.kind() != Kind.PROGRAM) {
            ApplicationFailure.union(
                    UnionException.invalidCase(
                            new InvalidCaseError("program", this.kind.name())));
        }

        return this.invocationValue.getValue();
    }

    /**
     * The summation case.
     * @return The summation.
     */
    public Summation summation()
    {
        if (this.kind() != Kind.SUMMATION) {
            ApplicationFailure.union(
                    UnionException.invalidCase(
                            new InvalidCaseError("summation", this.kind.name())));
        }

        return this.summation.getValue();
    }



    // > Variable
    // ------------------------------------------------------------------------------------------

    /**
     * Get the variable name which is a unique identifier.
     * @return The variable name.
     */
    public String name()
    {
        return this.name.getValue();
    }


    public List<VariableReference> dependencies()
    {
        List<VariableReference> variableDependencies = new ArrayList<>();

        switch (this.kind.getValue())
        {
            case LITERAL:
                break;
            case PROGRAM:
                variableDependencies = this.invocation().variableDependencies();
                break;
            case SUMMATION:
                variableDependencies = this.summation().variableDependencies();
                break;
            default:
                ApplicationFailure.union(
                        UnionException.unknownVariant(
                                new UnknownVariantError(Kind.class.getName())));
        }

        return variableDependencies;
    }


    /**
     * The variable's tags.
     * @return The tag list.
     */
    public List<String> tags()
    {
        return Arrays.asList(this.tags.getValue());
    }


    // > State
    // ------------------------------------------------------------------------------------------

    // ** Kind
    // ------------------------------------------------------------------------------------------

    /**
     * Get the kind of number variable.
     * @return The number variable kind.
     */
    private Kind kind()
    {
        return this.kind.getValue();
    }



    // ** Value
    // ------------------------------------------------------------------------------------------

    /**
     * Set the number variable integer. value
     * @param newValue The integer value.
     */
    public void setValue(Integer newValue)
    {
        switch (this.kind.getValue())
        {
            case LITERAL:
                this.integerValue.setValue(newValue);
                this.onUpdate();
                break;
            case PROGRAM:
                // Do Nothing?
                //this.reactiveValue.setValue(newValue);
                break;
            case SUMMATION:
                // Do Nothing?
                break;
        }
    }


    /**
     * Get the number variable's integer value.
     * @return The integer value.
     */
    public Integer value()
           throws VariableException
    {
        switch (this.kind.getValue())
        {
            case LITERAL:
                return this.integerValue.getValue();
            case PROGRAM:
                return this.reactiveValue.value();
            case SUMMATION:
                return this.summation.getValue().value();
        }

        return null;
    }


    /**
     * Get the value string representation. If the value contains any dice rolls, then it appears
     * as a formula, otherwise it is just an integer string.
     * @return The value string.
     * @throws SummationException
     */
    public String valueString()
           throws VariableException
    {
        switch (this.kind())
        {
            case LITERAL:
                return this.value().toString();
            case PROGRAM:
                return this.value().toString();
            case SUMMATION:
                return this.summation.getValue().valueString();
            default:
                ApplicationFailure.union(
                        UnionException.unknownVariant(
                                new UnknownVariantError(Kind.class.getName())));
        }

        return "";
    }


    // ** Refinement
    // ------------------------------------------------------------------------------------------

    /**
     * Returns true if the variable has a refinement.
     * @return True if the variable has a refinement.
     */
    public boolean hasRefinement()
    {
        return this.refinementId != null;
    }


    /**
     * Get the refinement identifier for this variable.
     * @return The variable's refinement id, or null if there is none.
     */
    public RefinementId getRefinementId()
    {
        return this.refinementId.getValue();
    }


    // INTERNAL
    // ------------------------------------------------------------------------------------------

    private void initialize()
    {
        // [1] Create reaction value (if program variable)
        // --------------------------------------------------------------------------------------

        if (this.kind.getValue() == Kind.PROGRAM) {
            this.reactiveValue = new ReactiveValue<>(this.invocationValue.getValue(),
                                                     VariableType.NUMBER);
        }
        else {
            this.reactiveValue = null;
        }
    }


    // KIND
    // ------------------------------------------------------------------------------------------

    public enum Kind
    {
        LITERAL,
        PROGRAM,
        SUMMATION;


        public static Kind fromString(String kindString)
                      throws InvalidDataException
        {
            return EnumUtils.fromString(Kind.class, kindString);
        }


        public static Kind fromYaml(Yaml yaml)
                      throws YamlException
        {
            String kindString = yaml.getString();
            try {
                return Kind.fromString(kindString);
            } catch (InvalidDataException e) {
                throw YamlException.invalidEnum(new InvalidEnumError(kindString));
            }
        }


        public static Kind fromSQLValue(SQLValue sqlValue)
                      throws DatabaseException
        {
            String enumString = "";
            try {
                enumString = sqlValue.getText();
                Kind kind = Kind.fromString(enumString);
                return kind;
            } catch (InvalidDataException e) {
                throw DatabaseException.invalidEnum(
                        new com.kispoko.tome.util.database.error.InvalidEnumError(enumString));
            }
        }

    }



}

File: app/src/main/java/com/kispoko/tome/model/engine/variable/NumberVariable.kt

package com.kispoko.tome.model.engine.variable


import com.kispoko.tome.lib.Factory
import com.kispoko.tome.lib.functor.Comp
import com.kispoko.tome.lib.functor.Func
import com.kispoko.tome.lib.functor.Prim
import com.kispoko.tome.lib.model.Model
import com.kispoko.tome.model.engine.program.Invocation
import com.kispoko.tome.model.engine.summation.Summation
import com.kispoko.tome.model.engine.value.ValueReference
import effect.Err
import effect.effApply
import effect.effApply2
import lulo.document.*
import lulo.value.UnexpectedType
import lulo.value.UnknownCase
import lulo.value.ValueError
import lulo.value.ValueParser
import lulo.value.valueResult
import java.util.*



/**
 * Number Variable
 */
sealed class NumberVariableValue : Model
{

     companion object : Factory<NumberVariableValue>
    {
        override fun fromDocument(doc: SpecDoc): ValueParser<NumberVariableValue> = when (doc)
        {
            is DocDict -> when (doc.case())
            {
                "literal"   -> NumberVariableLiteralIntegerValue.fromDocument(doc)
                "variable"  -> NumberVariableVariableValue.fromDocument(doc)
                "program"   -> NumberVariableProgramValue.fromDocument(doc)
                "value"     -> NumberVariableValueValue.fromDocument(doc)
                "summation" -> NumberVariableSummationValue.fromDocument(doc)
                else        -> Err<ValueError, DocPath,NumberVariableValue>(
                                    UnknownCase(doc.case()), doc.path)
            }
            else       -> Err(UnexpectedType(DocType.DICT, docType(doc)), doc.path)
        }
    }

}


/**
 * Literal Value
 */
data class NumberVariableLiteralIntegerValue(
                            override val id : UUID,
                            val value : Func<Int>) : NumberVariableValue()
{

    companion object : Factory<NumberVariableValue>
    {
        override fun fromDocument(doc : SpecDoc)
                      : ValueParser<NumberVariableValue> = when (doc)
        {
            is DocDict -> effApply2(::NumberVariableLiteralIntegerValue,
                                    // Model Id
                                    valueResult(UUID.randomUUID()),
                                    // Value
                                    effApply(::Prim, doc.int("value")))
            else       -> Err(UnexpectedType(DocType.DICT, docType(doc)), doc.path)
        }
    }

    override fun onLoad() { }

}


/**
 * Variable Value
 */
data class NumberVariableVariableValue(
                            override val id : UUID,
                            val variableReference : Func<VariableReference>) : NumberVariableValue()
{

    companion object : Factory<NumberVariableValue>
    {
        override fun fromDocument(doc : SpecDoc)
                      : ValueParser<NumberVariableValue> = when (doc)
        {
            is DocDict -> effApply2(::NumberVariableVariableValue,
                                    // Model Id
                                    valueResult(UUID.randomUUID()),
                                    // Variable Reference
                                    doc.at("reference") ap {
                                        effApply(::Comp, VariableReference.fromDocument(it))
                                    })
            else       -> Err(UnexpectedType(DocType.DICT, docType(doc)), doc.path)
        }
    }

    override fun onLoad() { }

}


/**
 * Program Value
 */
data class NumberVariableProgramValue(
                            override val id : UUID,
                            val inovcation : Func<Invocation>) : NumberVariableValue()
{

    companion object : Factory<NumberVariableValue>
    {
        override fun fromDocument(doc : SpecDoc)
                      : ValueParser<NumberVariableValue> = when (doc)
        {
            is DocDict -> effApply2(::NumberVariableProgramValue,
                                    // Model Id
                                    valueResult(UUID.randomUUID()),
                                    // Variable Reference
                                    doc.at("invocation") ap {
                                        effApply(::Comp, Invocation.fromDocument(it))
                                    })
            else       -> Err(UnexpectedType(DocType.DICT, docType(doc)), doc.path)
        }
    }

    override fun onLoad() { }

}


/**
 * Program Value
 */
data class NumberVariableValueValue(
                            override val id : UUID,
                            val valueReference : Func<ValueReference>) : NumberVariableValue()
{

    companion object : Factory<NumberVariableValue>
    {
        override fun fromDocument(doc : SpecDoc)
                      : ValueParser<NumberVariableValue> = when (doc)
        {
            is DocDict -> effApply2(::NumberVariableValueValue,
                                    // Model Id
                                    valueResult(UUID.randomUUID()),
                                    // Variable Reference
                                    doc.at("reference") ap {
                                        effApply(::Comp, ValueReference.fromDocument(it))
                                    })
            else       -> Err(UnexpectedType(DocType.DICT, docType(doc)), doc.path)
        }
    }

    override fun onLoad() { }

}


/**
 * Summation Value
 */
data class NumberVariableSummationValue(
                            override val id : UUID,
                            val summation : Func<Summation>) : NumberVariableValue()
{

    companion object : Factory<NumberVariableValue>
    {
        override fun fromDocument(doc : SpecDoc)
                      : ValueParser<NumberVariableValue> = when (doc)
        {
            is DocDict -> effApply2(::NumberVariableSummationValue,
                                    // Model Id
                                    valueResult(UUID.randomUUID()),
                                    // Variable Reference
                                    doc.at("summation") ap {
                                        effApply(::Comp, Summation.fromDocument(it))
                                    })
            else       -> Err(UnexpectedType(DocType.DICT, docType(doc)), doc.path)
        }
    }

    override fun onLoad() { }

}


//
//
//    @Override
//    public List<VariableReference> dependencies()
//    {
//        List<VariableReference> variableDependencies = new ArrayList<>();
//
//        switch (this.kind.getValue())
//        {
//            case LITERAL:
//                break;
//            case VARIABLE:
//                break;
//            case PROGRAM:
//                variableDependencies = this.invocation().variableDependencies();
//                break;
//            case VALUE:
//                break;
//            case SUMMATION:
//                variableDependencies = this.summation().variableDependencies();
//                break;
//            default:
//                ApplicationFailure.union(
//                        UnionException.unknownVariant(
//                                new UnknownVariantError(Kind.class.getName())));
//        }
//
//        return variableDependencies;
//    }
//
//
//    /**
//     * The variable's tags.
//     * @return The tag list.
//     */
//    @Override
//    public List<String> tags()
//    {
//        return Arrays.asList(this.tags.getValue());
//    }
//
//
//    /**
//     * Get the value string representation. If the value contains any dice rolls, then it appears
//     * as a formula, otherwise it is just an integer string.
//     * @return The value string.
//     * @throws NullVariableException
//     */
//    public String valueString()
//           throws NullVariableException
//    {
//        switch (this.kind())
//        {
//            case LITERAL:
//                return this.value().toString();
//            case VARIABLE:
//                return this.value().toString();
//            case PROGRAM:
//                return this.value().toString();
//            case VALUE:
//                return this.value().toString();
//            case SUMMATION:
//                return this.summation().valueString();
//            default:
//                ApplicationFailure.union(
//                        UnionException.unknownVariant(
//                                new UnknownVariantError(Kind.class.getName())));
//        }
//
//        return "";
//    }
//
//
//    @Override
//    public void initialize()
//    {
//        // [1] Add to state
//        // --------------------------------------------------------------------------------------
//        this.addToState();
//
//        // [2] Save original name and label values in case namespaces changes multiple times
//        // --------------------------------------------------------------------------------------
//        this.originalName  = name();
//        this.originalLabel = label();
//    }
//
//
//    // > State
//    // ------------------------------------------------------------------------------------------
//
//    // ** Kind
//    // ------------------------------------------------------------------------------------------
//
//    /**
//     * Get the kind of number variable.
//     * @return The number variable kind.
//     */
//    public Kind kind()
//    {
//        return this.kind.getValue();
//    }
//
//
//    // ** Cases
//    // ------------------------------------------------------------------------------------------
//
//
//
//    // ** Value
//    // ------------------------------------------------------------------------------------------
//
//    /**
//     * Set the number variable integer. value
//     * @param newValue The integer value.
//     */
//    public void setValue(Integer newValue)
//    {
//        switch (this.kind.getValue())
//        {
//            case LITERAL:
//                this.literalValue.setValue(newValue);
//                this.onUpdate();
//                break;
//            case VARIABLE:
//                break;
//            case PROGRAM:
//                // Do Nothing?
//                //this.reactiveValue.setLiteralValue(newValue);
//                break;
//            case VALUE:
//                break;
//            case SUMMATION:
//                // Do Nothing?
//                break;
//        }
//    }
//
//
//    /**
//     * Get the number variable's integer value.
//     * @return The integer value.
//     */
//    public Integer value()
//           throws NullVariableException
//    {
//        switch (this.kind.getValue())
//        {
//            case LITERAL:
//                return this.literalValue.getValue();
//            case VARIABLE:
//                return referencedVariableValue();
//            case PROGRAM:
//                return this.reactiveValue.value();
//            case VALUE:
//                Dictionary dictionary = SheetManagerOld.currentSheet().engine().dictionary();
//                NumberValue numberValue = dictionary.numberValue(this.valueReference());
//                return numberValue.value();
//            case SUMMATION:
//                return this.summation().value();
//        }
//
//        throw new NullVariableException();
//    }
//
//
//    private Integer referencedVariableValue()
//            throws NullVariableException
//    {
//        if (!State.hasVariable(this.variableReference())) {
//            ApplicationFailure.variable(
//                    VariableException.undefinedVariable(
//                            new UndefinedVariableError(this.variableReference())));
//            throw new NullVariableException();
//        }
//
//        VariableUnion variableUnion = State.variableWithName(this.variableReference());
//
//        // Variable is wrong type, log error, and return as null variable exception
//        if (variableUnion.type() != VariableType.NUMBER) {
//            ApplicationFailure.variable(
//                    VariableException.unexpectedVariableType(
//                            new UnexpectedVariableTypeError(this.variableReference(),
//                                                            VariableType.NUMBER,
//                                                            variableUnion.type())));
//            throw new NullVariableException();
//        }
//
//        return variableUnion.numberVariable().value();
//    }
//
//
//    // INTERNAL
//    // ------------------------------------------------------------------------------------------
//
//    // ** Initialize
//    // ------------------------------------------------------------------------------------------
//
//    private void initializeNumberVariable()
//    {
//        // [1] Create reaction value (if program variable)
//        // --------------------------------------------------------------------------------------
//
//        if (this.kind.getValue() == Kind.PROGRAM) {
//            this.reactiveValue = new ReactiveValue<>(this.invocationValue.getValue(),
//                                                     VariableType.NUMBER);
//        }
//        else {
//            this.reactiveValue = null;
//        }
//    }
//
//
//    private void initializeFunctors()
//    {
//        // Name
//        this.name.setName("name");
//        this.name.setLabelId(R.string.variable_field_name_label);
//        this.name.setDescriptionId(R.string.variable_field_name_description);
//
//        // Label
//        this.label.setName("label");
//        this.label.setLabelId(R.string.variable_field_label_label);
//        this.label.setDescriptionId(R.string.variable_field_label_description);
//
//        // Description
//        this.description.setName("description");
//        this.description.setLabelId(R.string.variable_field_description_label);
//        this.description.setDescriptionId(R.string.variable_field_description_description);
//
//        // Is Namespaced
//        this.isNamespaced.setName("is_namespaced");
//        this.isNamespaced.setLabelId(R.string.variable_field_is_namespaced_label);
//        this.isNamespaced.setDescriptionId(R.string.variable_field_is_namespaced_description);
//
//        // Tags
//        this.tags.setName("tags");
//        this.tags.setLabelId(R.string.variable_field_tags_label);
//        this.tags.setDescriptionId(R.string.variable_field_tags_description);
//
//        // Kind
//        this.kind.setName("kind");
//        this.kind.setLabelId(R.string.variable_field_kind_label);
//        this.kind.setDescriptionId(R.string.variable_field_kind_description);
//
//        // Literal Value
//        this.literalValue.setName("kind_literal");
//        this.literalValue.setLabelId(R.string.number_variable_field_value_literal_label);
//        this.literalValue.setDescriptionId(R.string.number_variable_field_value_literal_description);
//        this.literalValue.caseOf("kind", "literal");
//
//        // Variable Value
//        this.variableReference.setName("kind_variable");
//        this.variableReference.setLabelId(R.string.number_variable_field_value_variable_label);
//        this.variableReference.setDescriptionId(R.string.number_variable_field_value_variable_description);
//        this.variableReference.caseOf("kind", "variable");
//
//        // Program Value
//        this.invocationValue.setName("kind_program");
//        this.invocationValue.setLabelId(R.string.number_variable_field_value_program_label);
//        this.invocationValue.setDescriptionId(R.string.number_variable_field_value_program_description);
//        this.invocationValue.caseOf("kind", "program");
//
//        // Value (From Value Set) Value
//        this.valueReference.setName("kind_value");
//        this.valueReference.setLabelId(R.string.number_variable_field_value_value_label);
//        this.valueReference.setDescriptionId(R.string.number_variable_field_value_value_description);
//        this.valueReference.caseOf("kind", "value");
//
//        // Summation Value
//        this.summation.setName("kind_summation");
//        this.summation.setLabelId(R.string.number_variable_field_value_summation_label);
//        this.summation.setDescriptionId(R.string.number_variable_field_value_summation_description);
//        this.summation.caseOf("kind", "summation");
//    }
//
//
//    // ** Variable State
//    // ------------------------------------------------------------------------------------------
//
//    /**
//     * Add any variables associated with the current value to the state.
//     */
//    private void addToState()
//    {
//        if (this.kind() != Kind.VALUE)
//            return;
//
//        Dictionary dictionary = SheetManagerOld.currentSheet().engine().dictionary();
//        dictionary.numberValue(this.valueReference()).addToState();
//    }
//
//
//    private void removeFromState()
//    {
//        if (this.kind() != Kind.VALUE)
//            return;
//
//        Dictionary dictionary = SheetManagerOld.currentSheet().engine().dictionary();
//        dictionary.numberValue(this.valueReference()).removeFromState();
//    }
//

